lib/internal/async_hooks.js

```diff
@@ -245,25 +245,32 @@ function newUid() {
   return ++async_id_fields[kAsyncIdCounter];
 }
 
-
 // Return the triggerAsyncId meant for the constructor calling it. It's up to
 // the user to safeguard this call and make sure it's zero'd out when the
 // constructor is complete.
-function initTriggerId() {
-  var triggerAsyncId = async_id_fields[kInitTriggerAsyncId];
-  // Reset value after it's been called so the next constructor doesn't
-  // inherit it by accident.
-  async_id_fields[kInitTriggerAsyncId] = 0;
-  if (triggerAsyncId <= 0)
-    triggerAsyncId = async_id_fields[kExecutionAsyncId];
-  return triggerAsyncId;
+function getDefaultTriggerAsyncId() {
+  var defaultTriggerAsyncId = async_id_fields[kDefaultTriggerAsyncId];
+  // If defaultTriggerAsyncId isn't set, use the executionAsyncId
+  if (defaultTriggerAsyncId < 0)
+    defaultTriggerAsyncId = async_id_fields[kExecutionAsyncId];
+  return defaultTriggerAsyncId;
 }
 
 
-function setInitTriggerId(triggerAsyncId) {
+function defaultTriggerAsyncIdScope(triggerAsyncId, block) {
```

Comment:
I would add an opaque parameter here that we pass back to `block(opaque)`, so that we can avoid to allocate closures where it was not needed before.